Added a more extensive model-sets example (#9283)
astropy/astropy: docs/modeling/example-fitting-model-sets.rst

Fitting Model Sets
==================

Astropy model sets let you fit the same (linear) model to lots of independent

> data sets. It does the looping for you, and can be significantly faster than

Reviewer: Can you provide some estimate of the savings?  Maybe just a reasonable example with rough timings?  After reading this, I wondered if this was a factor of 10% improvement, factor of 2, factor of 10, ...

Reviewer 2: I would not use the word `looping`. It actually solves a set of linear equations simultaneously to find the exact solution. This is the reason also why it doesn't work with non-linear models which require an iterative fitting and looping.

Author: Added a time estimate. Factor of more than 600 for the use case of a 10x500x500 cube on my MacBook.

Author: Revised to say that it solves the linear equations simultaneously.

doing the looping yourself. But getting the data into the right shape can be 
a bit tricky. 

In the example below, we create a 3D data cube where the first dimension is a ramp, 
for example as from non-destructive readouts of an IR detector. So each pixel has a 
depth along a time axis, and flux that results a total number of counts that is 
increasing with time. We will be fitting a 1D polynomial vs. time to estimate the 
flux in counts/second (the slope of the fit).

   .. plot::
       :include-source:

       import numpy as np
       import matplotlib.pyplot as plt
       from scipy import stats
       from astropy.modeling import models, fitting
       
       # Set up the shape of the image and create the time axis

>        DEPTH,WIDTH,HEIGHT=10,101,102 # Time is along the DEPTH axis

Reviewer 3: Not sure how I feel about using all-caps variable names in an example.

Also can the cube be made smaller, so it won't take much time to run? We have a time limit imposed by ReadTheDocs. (Disclaimer: I haven't benchmarked this example.)

Author: Not sure how I feel about the all caps either. I was sort of thinking of them as constants, even though this isn't a module.

Reviewer 3: Other examples do not all-caps the constants. FYI.

Reviewer: Definitely could be a smaller cube for the example.

Author: Made it 10x4x3. Also changed to lower case.

       t = np.arange(DEPTH,dtype=np.float64)*10. # e.g. readouts every 10 seconds
       
       # Make up a flux in each pixel 

>        fluxes = np.arange(1.*WIDTH*HEIGHT).reshape(HEIGHT,WIDTH)

Reviewer 3: ```suggestion
       fluxes = np.arange(WIDTH * HEIGHT, dtype=np.float).reshape(HEIGHT, WIDTH)
```

Author: Added plot labels, but didn't extend the plots to the future as that seemed like making the example more complicated than it needed to be. 

Ran flake8 and addressed all the suggestions.

       # Create the ramps by integrating the fluxes along the time steps

>        image = np.stack([fluxes*dt for dt in t],axis=0)

Reviewer 3: ```suggestion
       image = fluxes[np.newaxis, :, :] * t[:, np.newaxis, np.newaxis]
```

Timing:
```python
In [53]: %timeit np.stack([fluxes*dt for dt in t],axis=0)                       
143 µs ± 319 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [54]: %timeit fluxes[np.newaxis, :, :] * t[:, np.newaxis, np.newaxis]        
47.9 µs ± 278 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

Author: Clever.

Author: Adopted this.

       # Add some Gaussian noise to each sample
       image += stats.norm.rvs(0.,image*0.05,size=image.shape) # Add noise

>        print("The image shape is: ",image.shape)

Reviewer 3: The print output cannot be doctested the way it is currently set up. I recommend refactoring such that it can be doctested, or just remove all the `print()`.

Author: I couldn't figure out how to get printing to work from within a .plot directive. Do you have suggestions for how to refactor to get this to work?

Reviewer 3: Can we just outright remove the `print`s? If the plot renders, we know the example works, isn't that enough?

Reviewer 4: One way to have a narrative flow of code snippets and then have the plot without `:include-source:`. That would unfortunately double the lines, but the rendered version will be much nicer with a logical flow and outputs that can be doctested.

Author: Refactored along the lines. I don't know how to tell if the tests are running and passing on the code snippets though. Also, it would be nice to give the user the option of downloading the entire source rather than the snippets, and I'm not sure how to do that.

Reviewer 4: Yes, as I recall we don't have a solution for downloading the figure source codes yet. Maybe this can be another way to extend the concept of examples? 
cc @kelle @jonathansick

       
       # Create the models and the fitter
       N=WIDTH*HEIGHT # This is how many instances we need
       line=models.Polynomial1D(degree=1,n_models=N)
       fit = fitting.LinearLSQFitter()
       print("We created %d models" % len(line))
       

>        # We need to get the data to be fit into the right shape

Reviewer: It it required to use model sets to have a 2D array instead of fitting the 3D array directly?  If so, it would be good to emphasize this in the text of the docs not just here in a comment.

Author: Yes. I've addressed it in the text now.

       # In this case, the time axis can be one dimensional.
       # The fluxes have to be organized into an array 
       # that is of shape `(WIDTH*HEIGHT,DEPTH)`
       # i.e we are reshaping to flatten last two axes and 
       # transposing to put them first.
       pixels = image.reshape((DEPTH,WIDTH*HEIGHT))
       y = pixels.T
       print("x axis is one dimensional: ",t.shape)
       print("y axis is two dimensional, N by len(x): ", y.shape)
       
       # Fit the model. It does the looping over the N models implicitly
       new_model = fit(line,x=t,y=y)
       print("We fit %d models: " % len(new_model))
       
       # Fill an array with values computed from the best fit and reshape it to match the original
       best_fit = new_model(t,model_set_axis=False).T.reshape((DEPTH,HEIGHT,WIDTH))
       print("We reshaped the best fit to dimensions: ",best_fit.shape)
       
       # Plot the fit along a couple of pixels
       plt.figure(figsize=(10,5))
       plt.subplot(121)

>        X,Y=HEIGHT//5,WIDTH//5

Reviewer 3: What does 5 mean?

Author: Yeah, that's probably too obscure. I was just trying to make a plot for a couple random pixels within the array. I'll do it differently.

Author: Just hardcoded in the pixel values; this is an example after all.

       plt.plot(t,image[:,X,Y],'.')
       plt.plot(t,best_fit[:,X,Y],'-')
       plt.subplot(122)

>        X,Y=HEIGHT//2,WIDTH//2

Reviewer 3: What does 2 mean?

Author: responded as for previous comment.

    >>> image = fluxes[np.newaxis, :, :] * t[:, np.newaxis, np.newaxis]
    >>> image += stats.norm.rvs(0., image*0.05, size=image.shape)  # Add noise
    >>> image.shape
    (10, 3, 4)

Create the models and the fitter. We need N=width*height instances of the same linear,
parametric model (model sets currently only work with linear models and fitters)::

    >>> N = width*height 
    >>> line = models.Polynomial1D(degree=1, n_models=N)
    >>> fit = fitting.LinearLSQFitter()
    >>> print("We created %d models" % len(line))
    We created 12 models

We need to get the data to be fit into the right shape. It's not possible to just feed
the 3D data cube. In this case, the time axis can be one dimensional. 
The fluxes have to be organized into an array that is of shape `width*height,depth` --  in 
other words, we are reshaping to flatten last two axes and transposing to put them first::

    >>> pixels = image.reshape((depth, width*height))
    >>> y = pixels.T
    >>> print("x axis is one dimensional: ",t.shape)
    >>> print("y axis is two dimensional, N by len(x): ", y.shape)
    x axis is one dimensional:  (10,)
    y axis is two dimensional, N by len(x):  (12, 10)

> Fit the model. It does the looping over the N models implicitly::

Reviewer 2: Remove the sentence with `looping`.

Author: Reworded

    >>> from astropy.modeling import models, fitting

We will use just a small image of 3 rows by 4 columns, with a depth of 10 non-destructive reads::

    >>> depth, width, height = 10, 3, 4  # Time is along the depth axis
    >>> t = np.arange(depth, dtype=np.float64)*10.  # e.g. readouts every 10 seconds

The number of counts in neach pixel is flux*time with the addition of some Gaussian noise::

    >>> fluxes = np.arange(1.*width*height).reshape(height, width)
    >>> image = fluxes[np.newaxis, :, :] * t[:, np.newaxis, np.newaxis]
    >>> image += stats.norm.rvs(0., image*0.05, size=image.shape)  # Add noise
    >>> image.shape
    (10, 3, 4)

Create the models and the fitter. We need N=width*height instances of the same linear,
parametric model (model sets currently only work with linear models and fitters)::

    >>> N = width*height 
    >>> line = models.Polynomial1D(degree=1, n_models=N)
    >>> fit = fitting.LinearLSQFitter()
    >>> print("We created %d models" % len(line))
    We created 12 models

We need to get the data to be fit into the right shape. It's not possible to just feed
the 3D data cube. In this case, the time axis can be one dimensional. 

> The fluxes have to be organized into an array that is of shape `width*height,depth` --  in 

Reviewer 2: @hcferguson because `width*height,depth` i sin single back ticks it attempts to link and fails. Use double back ticks here, 

```
 ``width*height,depth``
```

> The fluxes have to be organized into an array that is of shape `width*height,depth` --  in 

Reviewer 2: ```suggestion
The fluxes have to be organized into an array that is of shape ``width*height,depth`` --  in 
```

other words, we are reshaping to flatten last two axes and transposing to put them first::

    >>> pixels = image.reshape((depth, width*height))
    >>> y = pixels.T
    >>> print("x axis is one dimensional: ",t.shape)
    >>> print("y axis is two dimensional, N by len(x): ", y.shape)
    x axis is one dimensional:  (10,)
    y axis is two dimensional, N by len(x):  (12, 10)

> Fit the model. It does the looping over the N models implicitly::

Reviewer 2: And there's another loo[ping reference here.

> Fit the model. It does the looping over the N models implicitly::

Reviewer 2: ```suggestion
Fit the model. It fits the N models simultaneously::
```

    >>> new_model = fit(line, x=t, y=y)
    >>> print("We fit %d models" % len(new_model))
    We fit 12 models

Fill an array with values computed from the best fit and reshape it to match the original::

    >>> best_fit = new_model(t, model_set_axis=False).T.reshape((depth, height, width))

>     print("We reshaped the best fit to dimensions: ", best_fit.shape)

Reviewer 3: ```suggestion
    >>> print("We reshaped the best fit to dimensions: ", best_fit.shape)
```
